cache/secondary_cache_adapter.cc

```diff
@@ -236,13 +238,31 @@ Status CacheWithSecondaryAdapter::Insert(const Slice& key, ObjectPtr value,
                                          const Slice& compressed_value,
                                          CompressionType type) {
   Status s = target_->Insert(key, value, helper, charge, handle, priority);
-  if (s.ok() && value == nullptr && distribute_cache_res_) {
-    size_t sec_charge = static_cast<size_t>(
-        charge * (sec_cache_res_ratio_.load(std::memory_order_relaxed)));
-    s = secondary_cache_->Deflate(sec_charge);
-    assert(s.ok());
-    s = pri_cache_res_->UpdateCacheReservation(sec_charge, /*increase=*/false);
-    assert(s.ok());
+  if (s.ok() && value == nullptr && distribute_cache_res_ && handle) {
+    charge = target_->GetCharge(*handle);
+
+    MutexLock l(&mutex_);
+    placeholder_usage_ += charge;
+    // Check if total placeholder reservation is more than the overall
+    // cache capacity. If it is, then we don't try to charge the
+    // secondary cache because we don't want to overcharge it (beyond
+    // its capacity).
+    // In order to make this a bit more lightweight, we also check if
+    // the difference between placeholder_usage_ and reserved_usage_ is
+    // atleast kReservationChunkSize and avoid any adjustments if not.
+    if ((placeholder_usage_ <= target_->GetCapacity()) &&
+        ((placeholder_usage_ - reserved_usage_) >= kReservationChunkSize)) {
+      reserved_usage_ = placeholder_usage_ & ~(kReservationChunkSize - 1);
+      size_t new_sec_reserved =
+          static_cast<size_t>(reserved_usage_ * sec_cache_res_ratio_);
+      size_t sec_charge = new_sec_reserved - sec_reserved_;
+      s = secondary_cache_->Deflate(sec_charge);
```

Comment:
I'm not 100% convinced that there's no recursion in {`Insert`, `Release`} -> {`Deflate`, `UpdateCacheReservation`} -> {...}  that could lead to double-lock (usually deadlock) on `mutex_`. Can you convince me?

Reply:
The locks are acquired in the following order for both `Insert` and `Release` -

`mutex_` --> `pri_cache_res_->cache_res_mgr_mu_` --> `target_->mutex_` 

`mutex_` --> `secondary_cache_->cache_res_mgr_->cache_res_mgr_mu_` --> `secondary_cache_->cache_->mutex_`

So there's no recursion.

Reply:
I guess the only reference back to the `CacheWithSecondaryAdapter` from the primary cache (or secondary or `pri_cache_res_`) is the `SetEvictionCallback()` to `EvictionHandler()`. So `EvictionHandler()` could (probably?) be called while holding the new mutex but it doesn't call Insert or Release on itself.

Reply:
Good point. I forgot about the eviction handler. It might be called during `Release` while holding the new mutex, because we call `pri_cache_res_` to increase the reservation, which might lead to evictions. But like you said, it doesn't call Insert/Release on itself.

I'll admit it is quite convoluted. I'll add some asserts and rename the mutex to make its purpose clear so it doesn't accidentally get used in some other codepath.